patchutils.go

```diff
@@ -50,27 +59,41 @@ Loop:
 				continue Loop
 			case oldFileDiffs[i].NewName == "":
 				// File was deleted in old version
-				result += fmt.Sprintf("Only in %s: %s\n", filepath.Dir(newFileDiffs[j].NewName),
+				resultFiles[newFileDiffs[j].OrigName] = fmt.Sprintf("Only in %s: %s\n", filepath.Dir(newFileDiffs[j].NewName),
 					filepath.Base(newFileDiffs[j].NewName))
 			case newFileDiffs[j].NewName == "":
 				// File deleted in new version
-				result += fmt.Sprintf("Only in %s: %s\n", filepath.Dir(oldFileDiffs[i].NewName),
+				resultFiles[oldFileDiffs[i].OrigName] = fmt.Sprintf("Only in %s: %s\n", filepath.Dir(oldFileDiffs[i].NewName),
 					filepath.Base(oldFileDiffs[i].NewName))
 			default:
 				// interdiff of two versions
-				interFileDiff, err := interFileDiff(oldFileDiffs[i], newFileDiffs[j])
-
-				if err != nil {
-					return "", fmt.Errorf("merging diffs for file %q: %w", oldFileDiffs[i].OrigName, err)
-				}
-
-				fileDiffContent, err := diff.PrintFileDiff(interFileDiff)
-
-				if err != nil {
-					return "", fmt.Errorf("printing merged diffs for file %q: %w", oldFileDiffs[i].OrigName, err)
-				}
+				go func(oldD, newD *diff.FileDiff, ch chan interdiffResult) {
```

Comment:
Use golang.org/x/sync/errgroup instead. It takes care of waiting for all routines to finish. It's similar to sync.WaitGroup but adds error collection from each routine.

That'll allow you to get rid of the channel of structs and the collection and sorting of results at the end. To make that code safe, you'll need to guard all writes to the resultFiles map with a sync.Mutex.